File: nbcamp-java-todo-example/src/main/java/com/navi/nbcampjavatodoexample/api/service/TodoService.java
package com.navi.nbcampjavatodotask.api.service;

import com.navi.nbcampjavatodotask.api.model.exception.TodoNotFoundException;
import com.navi.nbcampjavatodotask.api.model.exception.TodoPasswordNotMatchedException;
import com.navi.nbcampjavatodotask.database.entity.Todo;
import com.navi.nbcampjavatodotask.database.repository.TodoRepository;
import java.time.LocalDateTime;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class TodoService {

    private final TodoRepository todoRepository;

    public Todo createTodo(
        String title,
        String content,
        String username,
        String password,
        LocalDateTime createdAt
    ) {
        Todo todo = Todo.builder()
            .title(title)
            .content(content)
            .username(username)
            .password(password)
            .createdAt(createdAt)
            .build();

        return todoRepository.save(todo);
    }

    public List<Todo> getTodos() {
        return todoRepository.findAllByOrderByCreatedAtDesc();
    }

    public Todo getTodoById(Long id) {
        return todoRepository.findById(id).orElseThrow(TodoNotFoundException::new);
    }

    public Todo updateTodo(
        Long id,
        String title,
        String content,
        String username,
        String password
    ) {
        Todo todo = getTodoById(id);

        if (!todo.getPassword().equals(password)) {
            throw new TodoPasswordNotMatchedException();
        }

        todo.update(title, content, username);

        return todoRepository.save(todo);
    }

    public void deleteTodo(
        Long id,
        String password
    ) {

        Todo todo = getTodoById(id);

        if (!todo.getPassword().equals(password)) {
            throw new TodoPasswordNotMatchedException();
        }

        todoRepository.delete(todo);
    }
}

File: nbcamp-kotlin-crud-example/src/main/kotlin/com/navi/nbcampkotlincrudexample/api/service/TodoService.kt
package com.navi.nbcampkotlincrudpractice.api.service

import com.navi.nbcampkotlincrudpractice.database.entity.Todo
import com.navi.nbcampkotlincrudpractice.database.repository.TodoRepository
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import java.time.LocalDateTime

@Service
class TodoService(
    private val todoRepository: TodoRepository,
    private val memberService: MemberService,
) {
    fun create(
        title: String,
        content: String,
        createdById: Long,
        createdAt: LocalDateTime,
    ): Todo {
        val createdBy = memberService.findById(createdById)

        return todoRepository.save(
            Todo(
                title = title,
                content = content,
                completed = false,
                createdBy = createdBy,
                createdAt = createdAt,
            ),
        )
    }

    fun findAll(): List<Todo> = todoRepository.findAll()

    fun findById(id: Long): Todo = todoRepository.findByIdOrNull(id) ?: throw IllegalStateException("Todo not found.")
}
